### ISSUE TITLE
Module.onRuntimeInitialized stops firing when compiled with pthreads and nonzero PTHREAD_POOL_SIZE

### ISSUE THREAD

**Issue by mewalig** [emscripten-core/emscripten#15730]
Hi,

I can successfully compile and load a module that contains a single exported function, which, depending on the arguments passed to it, may or may not require more than one thread.

If I compile without pthreads / PTHREAD_POOL_SIZE, it works fine so long as only the single-threaded features are used. When multiple threads are required, as expected, it hangs.

So, I recompiled with `-pthread` and linked with `-s USE_PTHREADS=1`. After doing that, when the multi-threaded features were attempted, I got an error about no unused workers available and having to set PTHREAD_POOL_SIZE.

So, I re-linked with `PTHREAD_POOL_SIZE=n`, and tried with n = 1, 2, 4, 8 and 16. In each case, I can now do *nothing*-- not even the single-threaded functions.

My console outputs the following every few seconds:
```
still waiting on run dependencies: 
dependency: wasm-instantiate
```

In looking at the generated js code further. I can see that unlike when it was compiled without pthreads and PTHREAD_POOL_SIZE, `Module.onRuntimeInitialized` is no longer firing.

On further review, I can see that `addRunDependency("wasm-instantiate")` gets called, but `removeRunDependency` never gets called, and going back further, I can see that in `loadWasmModuleToWorker`, `worker.onMessage` never fires (and therefore cannot call `removeRunDependency`), even though `worker.postMessage` does get invoked after `worker.onMessage` is properly set. 

I also tried changing `EXPORT_NAME` since there was some mention of that as well, all to no avail.

Any suggestions? This is frustrating, to say the least.

**Comment by sbc100** [comment]
I imagine this must be related to some other linker settings you are using since we have fairly wide coverage of pthread in the test suite.

Can you list all the settings you are using?  Since you mention `EXPORT_NAME`, can you trying building without `EXPORT_NAME` or `MODULARIZE`?

**Comment by mewalig** [comment]
Will do. just to confirm, onRuntimeInitialized should always fire, regardless of compilation / linker settings-- is that correct?

**Comment by sbc100** [comment]
Yes, its would always be a bug if `onRuntimeInitialized` doesn't fire. 

But also there could be some build options that are not tested with threading today.

**Comment by mewalig** [comment]
OK so I just spend the entire day to rebuild and re-install everything from scratch (including libs that are very annoying to install with emcc such as libicu and libxml2)-- from source code to make sure we had a consistent build environment. Same problem: onRuntimeInitialized() fires when it is compiled without multiple threads, and never fires if compiled with multiple threads.

I'm hoping this turns out to be a really embarrassingly simple trivial fix of the *doh!* variety, but at this point I'm out of ideas.

How exactly would you like me to list the settings I'm using?

As a start, I'll start with the basics:

javascript wrapper (call it `wrapper.js`):
```
var my_object = (function(Module){
    <xx.em.js contents go here: function GROWABLE_HEAP_I8() ... run()>

  Module.onRuntimeInitialized = function() {
    console.log('this code only runs when the wasm code was compiled without multiple threads');
  };

  return {
    invoke_my_c_code: function() { _my_exported_c_function(...); }
  };
})({
  print: function(x) { console.log('my print', x); },
  printErr: function(x) { console.log('my printerr', x); }
});
```

html:
```
<head>
  <script src="wrapper.js"></script>
</head>

```

http server:
```
#!/usr/bin/env python3
import http.server
import socket
import ssl
import os

server_address = ('0.0.0.0', 8888)

class CORSHTTPRequestHandler(http.server.SimpleHTTPRequestHandler):
    extensions_map = {
        '': 'application/octet-stream',
        '.manifest': 'text/cache-manifest',
        '.html': 'text/html',
        '.png': 'image/png',
        '.jpg': 'image/jpg',
        '.svg': 'image/svg+xml',
        '.css': 'text/css',
        '.js': 'application/x-javascript',
        '.wasm': 'application/wasm',
        '.json': 'application/json',
        '.xml': 'application/xml',
    }

    def end_headers(self):
        # Include additional response headers here. CORS for example:
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Cross-Origin-Opener-Policy', 'same-origin')
        self.send_header('Cross-Origin-Embedder-Policy', 'require-corp')
        http.server.SimpleHTTPRequestHandler.end_headers(self)

httpd = http.server.HTTPServer(server_address, CORSHTTPRequestHandler)
# ctx = ssl.SSLContext(ssl.PROTOCOL_TLS_SERVER)
# ctx.check_hostname = False
# ctx.load_cert_chain(certfile='localhost.pem')  # with key inside
# httpd.socket = ctx.wrap_socket(httpd.socket, server_side=True)
httpd.serve_forever()
```

**Comment by sbc100** [comment]
Thanks for the repro instructions.  I'll try an see it I can reproduce tomorrow.

A couple questions/pointers.   

- Is there some reason you specify the `onRuntimeInitialized` override in a different way to the `print` and `printErr` ones?  
- You should be able to use `emrun` to serve your application rather then a custom script.
- You can specify the `print`, `printErr`, and `onRuntimeInitialized` functions on the emcc command line using `--pre-js` file. 

None of these are necessary of course but can make repro cases smaller/simple and easier to share.

**Comment by mewalig** [comment]
Thank you. Further comments:

> You should be able to use `emrun`

Thank you, did not know. Much easier!

> `onRuntimeInitialized` vs `print` and `printErr`

For some reason, doing it the same way (overriding `print` and `printErr` the same way as `onRuntimeInitialized`) was not working before-- our assumption at the time was that, as seemingly hinted at in the documentation, `print` and `printErr` had to be overridden earlier in the initialization process). Though, that was several versions ago, and with an older version of `emcc`, so I'm not sure if that problem would still exist now

> using `--pre-js`

Agreed, that is the better / proper way to do all of this. No good reason here other than that we weren't aware of it at the time that we initially we initially began our project, and the current way was working fine (at least until we tried to add multi-threaded features), and we haven't yet had a chance to refactor the build script to use `--pre-js`, and we're trying to only solve for one moving part at a time. For now, that moving part is the multi-threading build process, but if using `--pre-js` might impact that, then happy to move that improvement to the forefront

**Comment by terop2** [comment]
oh, forgot to mention, that there's clear dependency to missing -g option on the onRuntimeInitialized. Thus this problem happen when you're trying to optimize the binary size and remove debug info from the binaries.

**Comment by sbc100** [comment]
I tried many ways to reproduce this with the latest emscripten.     If anyone is still able to reproduce please share clear reproduction steps.    I tried to reproduce using the wapper.js method described in https://github.com/emscripten-core/emscripten/issues/15730#issuecomment-990635035 but was unable to reprocuce.

Please re-open if this is still an issue.
